test/databases/transaction_database.cpp

```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
+    BOOST_REQUIRE(db.create());
+
+    // tx1: coinbase transaction
+    const chain::input::list tx1_inputs
+    {
+        { chain::point{ null_hash, chain::point::null_index }, {}, 0 }
+    };
+    
+    chain::output::list tx1_outputs;
+    tx1_outputs.emplace_back();
+    tx1_outputs.back().set_value(1200);
+
+    chain::transaction tx1(version, locktime, tx1_inputs, tx1_outputs);
+    BOOST_REQUIRE(tx1.is_coinbase());    
+    const auto hash1 = tx1.hash();
+    db.store(tx1, 1);
+    
+    const auto result1 = db.get(hash1);
+    BOOST_REQUIRE(result1);
+    BOOST_REQUIRE(result1.transaction().hash() == hash1);
+
+    // tx2: spends coinbase, tx1 as input, dummy output
+    chain::input::list tx2_inputs;
+    chain::output_point previous_output{hash1, 0};
+    chain::input instance(previous_output, {}, 0);
+    tx2_inputs.emplace_back(instance);
+
+    chain::output::list tx2_outputs;
+    tx2_outputs.emplace_back();
+    tx2_outputs.back().set_value(1200);
+    
+    chain::transaction tx2(version, locktime, tx2_inputs, tx2_outputs);
+    const auto hash2 = tx2.hash();
+    db.store(tx2, 1);
+
+    BOOST_REQUIRE(!result1.candidate());
+    BOOST_REQUIRE(!result1.transaction().outputs().front().metadata.candidate_spend);
+    
+    // Setup end    
+
+    const bool result = db.candidate(db.get(hash2).link());
+    BOOST_REQUIRE(result);
+    
+    const auto tx2_reloaded = db.get(hash2);
+    BOOST_REQUIRE(tx2_reloaded.candidate());
+
+    const auto tx1_reloaded = db.get(hash1);
+    BOOST_REQUIRE(!tx1_reloaded.candidate());
+    BOOST_REQUIRE(tx1_reloaded.transaction().outputs().front().metadata.candidate_spend);
+    
+}
+
+BOOST_AUTO_TEST_CASE(transaction_database__uncandidate__with_no_input_in_db__returns_false)
+{
+    transaction tx1;
+    data_chunk wire_tx1;
+    BOOST_REQUIRE(decode_base16(wire_tx1, TRANSACTION1));
+    BOOST_REQUIRE(tx1.from_data(wire_tx1));
+
+    const auto path = DIRECTORY "/tx_table";
+    test::create(path);
+    transaction_database db(path, 1000, 50, 0);
+    BOOST_REQUIRE(db.create());
+
+    const auto hash1 = tx1.hash();
+    db.store(tx1, 1);
+
+    const auto result1 = db.get(hash1);
+    BOOST_REQUIRE(result1);
+    BOOST_REQUIRE(result1.transaction().hash() == hash1);
+    
+    // Setup end    
+
+    const bool result = db.uncandidate(db.get(hash1).link());
+    BOOST_REQUIRE(!result);
+
+}
+
+BOOST_AUTO_TEST_CASE(transaction_database__uncandidate__with_input_in_db__returns_true)
+{
+    uint32_t version = 2345u;
+    uint32_t locktime = 0xffffffff;
+    
+    const auto path = DIRECTORY "/tx_table";
+    test::create(path);
+    transaction_database db(path, 1000, 50, 0);
+    BOOST_REQUIRE(db.create());
+
+    // tx1: coinbase transaction
+    const chain::input::list tx1_inputs
+    {
+        { chain::point{ null_hash, chain::point::null_index }, {}, 0 }
+    };
+    
+    chain::output::list tx1_outputs;
+    tx1_outputs.emplace_back();
+    tx1_outputs.back().set_value(1200);
```

Comment:
Avoid this construct, use initializer list instead (as with `tx1_inputs` above, or for the entire `tx1` (below)).

Reply:
The reason we want to avoid updating writable members is that it breaks the caching of derived property values. The exposure of writable properties is effectively deprecated.

Reply:
Switched to initializers everywhere I create inputs and outputs.